File: sketch/src/main/java/me/xiaopan/sketch/display/FadeInImageDisplayer.java
package me.xiaopan.sketch.display;

import android.graphics.drawable.Drawable;
import android.view.animation.AlphaAnimation;
import android.view.animation.DecelerateInterpolator;

import me.xiaopan.sketch.SketchView;

/**
 * 渐入动画
 */
@SuppressWarnings("unused")
public class FadeInImageDisplayer implements ImageDisplayer {
    private static final String KEY = "FadeInImageDisplayer";

    private int duration;
    private boolean alwaysUse;

    public FadeInImageDisplayer(int duration, boolean alwaysUse) {
        this.duration = duration;
        this.alwaysUse = alwaysUse;
    }

    public FadeInImageDisplayer(int duration) {
        this(duration, false);
    }

    public FadeInImageDisplayer(boolean alwaysUse) {
        this(DEFAULT_ANIMATION_DURATION, alwaysUse);
    }

    public FadeInImageDisplayer() {
        this(DEFAULT_ANIMATION_DURATION, false);
    }

    @Override
    public void display(SketchView sketchView, Drawable newDrawable) {
        if (newDrawable == null) {
            return;
        }
        AlphaAnimation animation = new AlphaAnimation(0.0f, 1.0f);
        animation.setInterpolator(new DecelerateInterpolator());
        animation.setDuration(duration);
        sketchView.clearAnimation();
        sketchView.setImageDrawable(newDrawable);
        sketchView.startAnimation(animation);
    }

    @Override
    public int getDuration() {
        return duration;
    }

    @Override
    public boolean isAlwaysUse() {
        return alwaysUse;
    }

    @Override
    public String getKey() {
        return String.format("%s(duration=%d, alwaysUse=%s)", KEY, duration, alwaysUse);
    }
}

File: sketch/src/main/java/com/github/panpf/sketch/display/FadeInImageDisplayer.kt
package com.github.panpf.sketch.display

import android.graphics.drawable.Drawable
import android.view.animation.AlphaAnimation
import android.view.animation.DecelerateInterpolator
import com.github.panpf.sketch.SketchView
import java.util.*

/**
 * 渐入动画
 */
class FadeInImageDisplayer @JvmOverloads constructor(
    override val duration: Int = ImageDisplayer.DEFAULT_ANIMATION_DURATION,
    override val isAlwaysUse: Boolean = false
) : ImageDisplayer {

    constructor(alwaysUse: Boolean) : this(ImageDisplayer.DEFAULT_ANIMATION_DURATION, alwaysUse)

    override fun display(sketchView: SketchView, newDrawable: Drawable) {
        sketchView.apply {
            clearAnimation()
            setImageDrawable(newDrawable)
            startAnimation(AlphaAnimation(0.0f, 1.0f).apply {
                interpolator = DecelerateInterpolator()
                duration = this@FadeInImageDisplayer.duration.toLong()
            })
        }
    }

    override fun toString(): String = "%s(duration=%d,alwaysUse=%s)"
        .format(Locale.US, "FadeInImageDisplayer", duration, isAlwaysUse)
}